packages/protocol/contracts/L1/libs/LibProving.sol

```diff
@@ -6,372 +6,188 @@
-        IProofVerifier proofVerifier,
-        TaikoData.Evidence memory evidence,
-        TaikoData.BlockMetadata memory target,
-        bytes32 blockHashOverride
-    ) private {
-        if (evidence.meta.id != target.id) revert L1_ID();
-        if (evidence.prover == address(0)) revert L1_PROVER();
+        TaikoData.BlockEvidence memory evidence
+    ) internal {
+        TaikoData.BlockMetadata memory meta = evidence.meta;
+        if (
+            meta.id != blockId ||
+            meta.id <= state.lastBlockId ||
+            meta.id >= state.nextBlockId
+        ) revert L1_ID();
 
-        if (FLAG_CHECK_METADATA) {
-            if (
-                target.id <= state.latestVerifiedId ||
-                target.id >= state.nextBlockId
-            ) revert L1_ID();
-            if (
-                state
-                    .getProposedBlock(config.maxNumBlocks, target.id)
-                    .metaHash != target.hashMetadata()
-            ) revert L1_META_MISMATCH();
-        }
+        if (
+            // 0 and 1 (placeholder) are not allowed
+            uint256(evidence.parentHash) <= 1 ||
+            // 0 and 1 (placeholder) are not allowed
+            uint256(evidence.blockHash) <= 1 ||
+            // cannot be the same hash
+            evidence.blockHash == evidence.parentHash ||
+            // 0 and 1 (placeholder) are not allowed
+            uint256(evidence.signalRoot) <= 1 ||
+            // prover must not be zero
+            evidence.prover == address(0)
+        ) revert L1_INVALID_EVIDENCE();
+
+        TaikoData.ProposedBlock storage proposal = state.blocks[
+            meta.id % config.maxNumBlocks
+        ];
 
-        if (FLAG_VALIDATE_HEADER_FOR_METADATA) {
-            if (
-                evidence.header.parentHash == 0 ||
-                evidence.header.beneficiary != evidence.meta.beneficiary ||
-                evidence.header.difficulty != 0 ||
-                evidence.header.gasLimit !=
-                evidence.meta.gasLimit + config.anchorTxGasLimit ||
-                evidence.header.gasUsed == 0 ||
-                evidence.header.timestamp != evidence.meta.timestamp ||
-                evidence.header.extraData.length !=
-                evidence.meta.extraData.length ||
-                keccak256(evidence.header.extraData) !=
-                keccak256(evidence.meta.extraData) ||
-                evidence.header.mixHash != evidence.meta.mixHash
-            ) revert L1_META_MISMATCH();
-        }
+        if (proposal.metaHash != LibUtils.hashMetadata(meta))
+            revert L1_EVIDENCE_MISMATCH();
 
-        // For alpha-2 testnet, the network allows any address to submit ZKP,
-        // but a special prover can skip ZKP verification if the ZKP is empty.
+        TaikoData.ForkChoice storage fc = state.forkChoices[
+            blockId % config.maxNumBlocks
+        ][proposal.nextForkChoiceId];
 
         bool oracleProving;
+        if (uint256(fc.chainData.blockHash) <= 1) {
```

Comment:
Isn't it always true ? Because `fc` is created like this:
```
TaikoData.ForkChoice storage fc = state.forkChoices[
            blockId % config.maxNumBlocks
        ][proposal.nextForkChoiceId];
```
and when this proveBlock() once succeeds, the proposed block's `nextForkChoiceId` incremented, so when some other proposer calls this to prove, it will create another, empty `fc` instance.

Reply:
- If oracle prover is disabled, then when a fork choice is proven by a regular prover, `fc.chainData.blockHash` will be 0 or 1 initially then changed to another value. Now if another prover tries to prove the same FC, the value is not 0 or 1 and the proving will fail.
- If oracle prover is enabled, the first prover must be the oracle prover, when a second, regular prover is trying to prove the block, `fc.chainData.blockHash` will not be 0 or 1.